File: 0121_best_time_to_buy_and_sell_stock.py
# # https://leetcode-cn.com/problems/best-time-to-buy-and-sell-stock/

class Solution:
    def maxProfit(self, prices):
        """
        Inputs:
            prices: list[int]
        Outputs:
            res: int
        """
        
        res = 0
        min_price = prices[0]
        for price in prices:
            # Find the minimum price till now.
            min_price = min(min_price, price)
            # Find the maximum profit till now.
            res = max(res, price-min_price)
        return res

File: Cpp/0121_best_time_to_buy_and_sell_stock.cpp
// https://leetcode.cn/problems/best-time-to-buy-and-sell-stock/

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int res = 0, min_stock = prices[0];
        for (int i = 1; i < prices.size(); ++i) {
            res = max(res, prices[i] - min_stock);
            min_stock = min(min_stock, prices[i]);
        }
        return res;
    }
};
